Shared Tree: In-memory format and API changes for persisted schema metadata

TommyBrosman opened the pull request:
Adds in-memory format and API changes for the persisted schema metadata feature.

# Changes

- Add metadata fields to Tree*StoredSchema types
  - Node
  - Field
- Add fields to in-memory format for persisted node/field schema metadata
  - Expose JsonCompatibleReadonlyObject as an alpha API so it can be used for the view schema format
- SchemaFactory changes
  - Introduce or update alpha types in places where persisted metadata will be passed: NodeSchemaOptionsAlpha, FieldPropsAlpha, FieldSchemaAlpha, FieldSchemaMetadataAlpha
  - Update internal code to use the alpha types
  - Updated node APIs in SchemaFactoryAlpha to pass `persistedMetadata` associated with nodes
  - Updated field APIs in SchemaFactoryAlpha to take new types for specifying `persistedMetadata` associated with fields
  - Test coverage in schemaFactory.spec.ts for the new node schema and field schema APIs
- simple-tree changes

# Notes for reviewers

## Todo

- [x] Merge with main
- [x] Remove redundant field factory methods
- [x] Revert files that belong in the FormatV2 PR
- [x] Update SchemaFactoryAlpha to use the changes in #24801

TommyBrosman committed:
Initial change.
- Added in-memory schema format v2.
- Implemented the schema format v2 codec.
- Updated test utils that previously assumed there was only one format.
- Added snapshots for many of the v2 tests.

TommyBrosman committed:
persistedMetadata is now backed by an object instead of a string.

TommyBrosman committed:
Updated codec.spec.ts to cover FormatV2.

TommyBrosman committed:
Removed extraneous whitespace change.

TommyBrosman committed:
- Changed TreeNodeStoredSchema implementations to take the write version as a parameter.
- Fixed naming of schema index formats.
- Changed schemaChangeFormat to be schema format-agnostic.

TommyBrosman committed:
- Changed the stored schema implementation to handle multiple schema formats.
- Moved SchemaCodecVersion.

TommyBrosman committed:
Merge branch 'main' into metadata-schema-3

TommyBrosman committed:
Apply suggestions from code review

Co-authored-by: Craig Macomber (Microsoft) <42876482+CraigMacomber@users.noreply.github.com>

TommyBrosman committed:
- Switched to using the SchemaCodecVersion in public APIs. Eventually we want something more intuitive (e.g., minimum client version).
- Split the encode method for TreeNodeStoredSchema.

TommyBrosman committed:
Merge branch 'metadata-schema-3' of https://github.com/TommyBrosman/FluidFramework into metadata-schema-3

TommyBrosman committed:
- Removed the schema version constant and replaced it with SchemaCodecVersion.
- Changed schema-stored/FormatV2 to only define the new elements.
- Updated all instantiations of TreeFieldStoredSchema so that persistedMetadata is specified.
- HACK: importCompressed needs to handle both schema formats. A few of the roundtrip JSON tests are failing.

TommyBrosman committed:
Minor: reverted changes to FluidClientVersion utils.

TommyBrosman committed:
Switched back to using min client version for importCompressed.

TommyBrosman committed:
Minor: reverted accidental change.

TommyBrosman committed:
Apply suggestions from code review

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

TommyBrosman committed:
Made persisted metadata schema field naming consistent.

TommyBrosman committed:
- Created a separate storedSchemaDecodeDispatcher for v2 schemas.
- Made importCompressed always use a v1 schema when encoding.

TommyBrosman committed:
- Refactored the node kind dispatch.
- Added better test data to one of the codec tests for the v2 schema format.

TommyBrosman committed:
Updated snapshots.

TommyBrosman committed:
- Removed minimum client version changes. Equivalent changes will be added in another PR.
- Updated a comment.

TommyBrosman committed:
Removed more minimumClientVersion glue.

TommyBrosman committed:
Updated API files.

TommyBrosman committed:
Rename: SchemaCodecVersion -> SchemaVersion.

TommyBrosman committed:
Merge branch 'main' into metadata-schema-3

TommyBrosman committed:
Removed top-level persisted metadata.

TommyBrosman committed:
Fixed comments.

TommyBrosman committed:
Exposed persistedMetadata as a JsonCompatibleReadOnlyObject.

TommyBrosman committed:
Changeset.

TommyBrosman committed:
Wired up toStoredSchema. Still needs tests.

TommyBrosman committed:
- Updated an old snapshot that didn't include the metadata field.
- Added tests calling schemaFactoryAlpha.

TommyBrosman committed:
Refactor: persistedMetadata -> metadata on persisted types.

TommyBrosman committed:
Apply suggestions from code review

Co-authored-by: Craig Macomber (Microsoft) <42876482+CraigMacomber@users.noreply.github.com>

TommyBrosman committed:
- Updated the changeset description
- Moved JsonCompatibleReadOnly and JsonCompatibleReadOnlyObject to alpha API status so they can be used in schemas
- Implemented field factory methods and added test coverage
- Added requiredRecursive to schemaFactoryAlpha (not sure if it should be part of this change though)

TommyBrosman committed:
Reverted unnecessary change.

TommyBrosman committed:
Wired up node schema metadata persistence. Currently errors on tests involving FormatV2.

TommyBrosman committed:
Reverted unneeded change.

TommyBrosman committed:
Removed tests for the alpha API I removed in a previous revision.

TommyBrosman committed:
This revision adds simple-tree persistence for node and field schema metadata.

- Fixed a bug where FormatV2 was including FormatV1 fields.
- persistedMetadata is now only written when non-undefined.
- simple-tree implementations for individual schema kinds (arrayNode, mapNode, objectNode) now pass persistedMetadata down from the outer factory calls.
- Introduced alpha APIs for simple-tree in places where we need to pass persistedMetadata but don't want to pollute the public API.
- Updated tests to pass persistedMetadata when constructing SimpleNodeSchemas.
- Wrote simple tests for node and field persisted metadata.

TommyBrosman committed:
- Fixed missing persistedMetadata on fields. toStoredSchema.ts and sharedTree.ts weren't copying them in all cases where the needed to, and I was missing SimpleFieldSchema.persistedMetadata.
- Added the v5 Shared Tree format and associated snapshots.
- Fixed FieldSchemaAlpha.persistedMetadata so it won't return an empty object if the data is undefined.

TommyBrosman committed:
Merge branch 'main' into metadata-schema-3

TommyBrosman committed:
Removed changes for persisted format, codecs. A few tests are failing.

TommyBrosman commented on .changeset/brown-dingos-switch.md:
```diff
@@ -0,0 +1,8 @@
+---
```
Todo: update.

TommyBrosman committed:
Minor cleanup.

TommyBrosman committed:
Updated changeset description.

TommyBrosman committed:
Merge branch 'main' into metadata-in-memory

TommyBrosman committed:
Regenerated API Extractor output.

alexvy86 commented on packages/dds/tree/src/simple-tree/toStoredSchema.ts:
```diff
@@ -90,7 +90,7 @@ export function convertField(schema: SimpleFieldSchema): TreeFieldStoredSchema {
 		convertFieldKind.get(schema.kind)?.identifier ?? fail(0xae3 /* Invalid field kind */);
 	const types: TreeTypeSet = schema.allowedTypesIdentifiers as TreeTypeSet;
-	return { kind, types };
+	return { kind, types, metadata: schema.persistedMetadata };
```
So if I'm getting this right, in the StoredSchema layer we're calling it plain `metadata` because it's the only one that we persist so `persistedMetadata` would be redundant (and as Craig pointed out at some point, wasteful because the property name consumes actual bytes on the wire and when stored at rest), and when moving to the SimpleSchema layer is when we map it into `persistedMetadata` to differentiate it from the transient one that already exists in the SimpleSchema domain, correct?

alexvy86 replied:
If so, maybe we could benefit from `@privateremarks` on the relevant properties on both sides to clarify it for people who start working in this part of the codebase.

alexvy86 commented on packages/dds/tree/src/test/simple-tree/api/getSimpleSchema.spec.ts:
```diff
@@ -24,12 +24,14 @@ const simpleString: SimpleLeafNodeSchema = {
 	leafKind: ValueSchema.Number,
 	kind: NodeKind.Leaf,
 	metadata: {},
+	persistedMetadata: undefined,
```
Seeing that the transient metadata is always initialized to `{}` makes me wonder if we should do the same for the persisted one for consistency. Make it required instead of optional at the simple schema layer and default it to an empty object when mapping an undefined from the stored schema layer.

TommyBrosman commented on packages/dds/tree/src/simple-tree/schemaTypes.ts:
```diff
@@ -774,17 +816,23 @@ function areFieldPropsEqual(a: FieldProps | undefined, b: FieldProps | undefined
 
-	return a?.custom === b?.custom && a?.description === b?.description;
+	return (
+		a?.custom === b?.custom &&
+		a?.description === b?.description &&
+		a?.persistedMetadata === b?.persistedMetadata
```
It is used to check equality of the FieldSchema. I see this comment on `areFieldPropsEqual` which is used by the FieldSchema comparison:

```
 * @remarks FieldProps are considered equivalent if their keys and default providers are reference equal, and their metadata are {@link areMetadataEqual | equivalent}.
 ```
 
I guess the word "equivalent" is kind of fuzzy. (Also: I need to update the comment on the metadata equality check). Per @CraigMacomber we want to use the Alpha type internally wherever we can. It doesn't make sense for two Alpha types to be considered equivalent if their persisted metadata is different. We have another backlog item for checking schema equivalence while ignoring persisted metadata (for older clients), so the comparison protocol may change at that point.

TommyBrosman replied:
Actually: should I even be checking reference equality here?

Josmithr replied:
We don't do deep equality comparison for the non-persisted metadata because there are no restrictions on what a user can provide in them. @noencke ran into some issues around this, as I recall.

For the persisted metadata, since we require and force that the data is serializable, it is probably more reasonable to do a deep comparison. Not sure what the perf considerations are for this path though.

TommyBrosman replied:
I discussed this more with @CraigMacomber and @Abe27342. For now it sounds like we are fine doing a shallow equality check (e.g., using stringify) that doesn't necessarily return "equivalent" in all cases where the persistedMetadata has the same contents. The only upstream caller for this method in non-test code is in schematizingTreeView.ts: hasRootSchema: [https://github.com/microsoft/FluidFramework/blob/c7ac860842f1a202c56ddf0573d28622ecb2fb2f/packages/…](https://github.com/microsoft/FluidFramework/blob/c7ac860842f1a202c56ddf0573d28622ecb2fb2f/packages/dds/tree/src/shared-tree/schematizingTreeView.ts#L161) However, this method is only called in test code. I'm going to create a work item to update equality checking.

Josmithr replied:
Probably worth adding a comment in the code as well.

TommyBrosman committed:
Renamed the in-memory version of the persistedMetadata field to persistedMetadata.

TommyBrosman committed:
Merge branch 'main' into metadata-in-memory

TommyBrosman commented on .changeset/brown-dingos-switch.md:
```diff
@@ -0,0 +1,8 @@
+---
+Add APIs for declaring "persisted" schema metadata without persistence
+
+Add alpha APIs for declaring node and field schema metadata which future versions of the Fluid Framework will provide a way to opt into persisting in the document. The new APIs function, but will not yet persist data.
```
The plan is to update the existing changeset with more information once persistence is implemented. @CraigMacomber pointed out that it doesn't make much sense to write a ton of documentation for an API that isn't meant to be used yet, so I kept it vague for now.

CraigMacomber replied:
Is there a reason we don't make these APIs internal instead of alpha (and have no changeset)? Getting things working end to end with everything being internal then promoting that to alpha with nice docs and a changeset seems like it would be preferable if practical.

TommyBrosman replied:
That's fair. I guess it might be more confusing to consume code with a changeset that doesn't apply to any real current consumer. I'll remove it for now.

TommyBrosman commented on packages/dds/tree/src/test/simple-tree/api/schemaFactory.spec.ts:
```diff
@@ -413,6 +413,56 @@ describe("schemaFactory", () => {
 		});
 
+		it("Node schema persisted metadata", () => {
+			const factory = new SchemaFactoryAlpha("");
```
I used the same pattern as `it("Node schema metadata" ...` above. I can change it.

TommyBrosman replied:
Fixed.

TommyBrosman committed:
- Added a simple default comparison for persistedMetadata.
- Updated a test (PR feeback).

CraigMacomber commented on packages/dds/tree/src/simple-tree/schemaTypes.ts:
```diff
@@ -295,6 +296,22 @@ export interface FieldProps<TCustomMetadata = unknown> {
+ * @typeParam TCustomMetadata - Custom metadata properties to associate with the field.
+ * See {@link FieldSchemaMetadata.custom}.
+ *
+ * @alpha
```
I believe the new input tag would apply here.
```suggestion
 * @alpha @input
```

TommyBrosman committed:
Minor: clarified use of persistedMetadata in tests.

TommyBrosman commented on packages/dds/tree/src/core/schema-stored/schema.ts:
```diff
@@ -25,6 +31,8 @@ export enum SchemaVersion {
 	v1 = 1,
 }
 
+type FieldSchemaFormat = FieldSchemaFormatV1;
```
Nice catch. I missed that.

TommyBrosman commented on packages/dds/tree/src/core/schema-stored/schema.ts:
```diff
@@ -194,7 +219,7 @@ export class ObjectNodeStoredSchema extends TreeNodeStoredSchema {
 	}
 
-	public override encode(): TreeNodeSchemaDataFormat {
+	public override encodeV1(): TreeNodeSchemaDataFormatV1 {
```
I think I'll add them back in the format PR. That would make more sense.

TommyBrosman committed:
- Removed redundant alpha APIs. For now they are exposed only as static
  methods, though that will change before the PR goes in.
- Removed the changeset. It is not needed as we are not yet releasing the feature (even as alpha) until the format changes land.

TommyBrosman commented on packages/dds/tree/src/simple-tree/api/schemaFactoryAlpha.ts:
```diff
@@ -157,16 +167,61 @@ export class SchemaFactoryAlpha<
+	/**
+	 * {@inheritDoc SchemaStatics.optional}
+	 */
+	public optionalAlpha<
```
I removed the redundant methods, will close this once I've updated SchemaFactoryAlpha to use the changes from your PR (#24801).

TommyBrosman committed:
Reverted changes that belong in a different PR.

TommyBrosman committed:
Removed more changes that shouldn't be in this PR.

TommyBrosman committed:
Apply suggestions from code review

Co-authored-by: Craig Macomber (Microsoft) <42876482+CraigMacomber@users.noreply.github.com>

TommyBrosman committed:
Reverted another unrelated change.

TommyBrosman committed:
Merge branch 'main' into metadata-in-memory

TommyBrosman committed:
Updated API Extractor output.

TommyBrosman committed:
Changed SchemaFactoryAlpha calls to use instance methods.

CraigMacomber commented on packages/dds/tree/api-report/tree.alpha.api.md:
```diff
@@ -596,6 +610,11 @@ export interface NodeSchemaOptions<out TCustomMetadata = unknown> {
 }
 
+// @alpha @sealed
+export interface NodeSchemaOptionsAlpha<out TCustomMetadata = unknown> extends NodeSchemaOptions<TCustomMetadata> {
```
It seems like users are suppose to create an instance of this to pass into your APIs? If so, marking it as sealed means they are not allowed to do that, and instead have to get the instance of it from one of our APIs. THat seems wrong.

Its use case seems more like the opposite, input, with gets the opposite compatability testing (variance wise).

CraigMacomber commented on packages/dds/tree/src/core/schema-stored/schema.ts:
```diff
@@ -116,6 +122,16 @@ export interface SchemaPolicy {
+ * @privateRemarks
+ * TODO: Move to schema FormatV2 once it has been defined.
+ */
+export const PersistedMetadataFormat = Type.Optional(JsonCompatibleReadOnlySchema);
```
I think you should just be using JsonCompatibleReadOnly below, rather than wrapping it up in TypeBox schema then unwrapping it. Generally the in memory representation should not use typebox types.

TommyBrosman replied:
I'm using JsonCompatibleReadOnlyObject in other places, so I'll use that here.

TommyBrosman committed:
- Cleaned up the type for in-memory persistedMetadata.
- Fixed an API annotation.
- Lint.

TommyBrosman committed:
Merge branch 'main' into metadata-in-memory

TommyBrosman committed:
Regenerated API Extractor output.

Josmithr commented on packages/dds/tree/src/simple-tree/node-kinds/array/arrayNode.ts:
```diff
@@ -1095,6 +1096,7 @@ export function arraySchema<
 	implicitlyConstructable: ImplicitlyConstructable,
 	customizable: boolean,
 	metadata?: NodeSchemaMetadata<TCustomMetadata>,
+	persistedMetadata?: JsonCompatibleReadOnlyObject | undefined,
```
(same for the other factory functions that take `persistedMetadata`)

TommyBrosman replied:
I documented `persistedMetadata` on all of the simple-tree object schema methods.

Josmithr reviewed (approved):
Approving for API changes. Would probably be good to get another tree signoff as well for the changes in the codec layers.

TommyBrosman committed:
Update packages/dds/tree/src/simple-tree/api/schemaFactory.ts

Co-authored-by: Joshua Smithrud <54606601+Josmithr@users.noreply.github.com>

TommyBrosman committed:
Docs.

TommyBrosman committed:
Merge branch 'metadata-in-memory' of https://github.com/TommyBrosman/FluidFramework into metadata-in-memory

TommyBrosman committed:
Merge branch 'main' into metadata-in-memory

CraigMacomber commented on packages/dds/tree/src/core/schema-stored/schema.ts:
```diff
@@ -128,6 +133,15 @@ export interface TreeFieldStoredSchema {
+	 * @remarks
+	 * Discarded when encoding to {@link SchemaFormatVersion.V1}.
+	 * @privateRemarks
+	 * This field corresponds to the `metadata` field in the persisted schema format.
```
You are referring to something that will exist as if it currently exists here.

CraigMacomber commented on packages/dds/tree/src/simple-tree/schemaTypes.ts:
```diff
@@ -775,17 +817,48 @@ function areFieldPropsEqual(a: FieldProps | undefined, b: FieldProps | undefined
+ * @remarks
+ * Currently only handles shallow equality in the case where the keys are in the same order. This is acceptable for current use cases.
+ */
+function arePersistedMetadataEqual(
```
This new functionality does not neem to have tests. Add cases for this in the test suite for this file (which tests this via the "areImplicitFieldSchemaEqual" test.)

CraigMacomber commented on packages/dds/tree/src/simple-tree/schemaTypes.ts:
```diff
@@ -775,17 +817,48 @@ function areFieldPropsEqual(a: FieldProps | undefined, b: FieldProps | undefined
+	}
+
+	return (
+		a?.custom === b?.custom &&
```
this should use Object.is since custom metadata could be something where === is not reflexive, like NaN. Also -0 and 0 are distinct in this context which === also gets wrong.

CraigMacomber commented on packages/dds/tree/src/simple-tree/schemaTypes.ts:
```diff
@@ -775,17 +817,48 @@ function areFieldPropsEqual(a: FieldProps | undefined, b: FieldProps | undefined
+	a: JsonCompatibleReadOnlyObject | undefined,
+	b: JsonCompatibleReadOnlyObject | undefined,
+): boolean {
 	if (a === b) {
```
I'd prefer use of Object.is here for consistency with cases where the difference matters (above) though in this case it doesn't actually matter since numbers are not allowed here.

TommyBrosman committed:
- Cleaned up comparators and added tests.
- Updated a doc comment.

TommyBrosman committed:
Merge branch 'main' into metadata-in-memory

TommyBrosman committed:
Merge branch 'main' into metadata-in-memory
